Userland/Libraries/LibAudio/PlaybackStreamPulseAudio.cpp

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (c) 2023, Gregory Bertilson <zaggy1024@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "PlaybackStreamPulseAudio.h"
+
+#include <LibCore/ThreadedPromise.h>
+
+namespace Audio {
+
+#define TRY_OR_EXIT_THREAD(expression)                                                                       \
+    ({                                                                                                       \
+        auto&& __temporary_result = (expression);                                                            \
+        if (__temporary_result.is_error()) [[unlikely]] {                                                    \
+            warnln("Failure in PulseAudio control thread: {}", __temporary_result.error().string_literal()); \
+            internal_state->exit();                                                                          \
+            return 1;                                                                                        \
+        }                                                                                                    \
+        __temporary_result.release_value();                                                                  \
+    })
+
+ErrorOr<NonnullRefPtr<PlaybackStream>> PlaybackStreamPulseAudio::create(OutputState initial_state, u32 sample_rate, u8 channels, u32 target_latency_ms, AudioDataRequestCallback&& data_request_callback)
+{
+    VERIFY(data_request_callback);
+
+    // Create an internal state for the control thread to hold on to.
+    auto internal_state = TRY(adopt_nonnull_ref_or_enomem(new (nothrow) InternalState()));
+    auto playback_stream = TRY(adopt_nonnull_ref_or_enomem(new (nothrow) PlaybackStreamPulseAudio(internal_state)));
+
+    // Create the control thread and start it.
+    auto thread = TRY(Threading::Thread::try_create([=, data_request_callback = move(data_request_callback)]() mutable {
+        auto context = TRY_OR_EXIT_THREAD(PulseAudioContext::instance());
+        internal_state->set_stream(TRY_OR_EXIT_THREAD(context->create_stream(initial_state, sample_rate, channels, target_latency_ms, [data_request_callback = move(data_request_callback)](PulseAudioStream&, Bytes buffer, size_t sample_count) {
+            return data_request_callback(buffer, PcmSampleFormat::Float32, sample_count);
+        })));
```

Comment:
I mentioned it on discord in #multimedia, but there's a race condition here. Some sites on my computer (like https://www.dofactory.com/html/audio ) end up trying to call some methods on the Audio element before this thread has had a chance to initialize the PulseAudioStream.

Two possible fixes:
1) call set_stream from the GUI thread
2) this patch 

```
diff --git a/Userland/Libraries/LibAudio/PlaybackStreamPulseAudio.cpp b/Userland/Libraries/LibAudio/PlaybackStreamPulseAudio.cpp
index d5cd374f56..0d1d4c68e6 100644
--- a/Userland/Libraries/LibAudio/PlaybackStreamPulseAudio.cpp
+++ b/Userland/Libraries/LibAudio/PlaybackStreamPulseAudio.cpp
@@ -29,24 +29,35 @@ ErrorOr<NonnullRefPtr<PlaybackStream>> PlaybackStreamPulseAudio::create(OutputSt
     auto internal_state = TRY(adopt_nonnull_ref_or_enomem(new (nothrow) InternalState()));
     auto playback_stream = TRY(adopt_nonnull_ref_or_enomem(new (nothrow) PlaybackStreamPulseAudio(internal_state)));
 
-    // Create the control thread and start it.
-    auto thread = TRY(Threading::Thread::try_create([=, data_request_callback = move(data_request_callback)]() mutable {
-        auto context = TRY_OR_EXIT_THREAD(PulseAudioContext::instance());
-        internal_state->set_stream(TRY_OR_EXIT_THREAD(context->create_stream(initial_state, sample_rate, channels, target_latency_ms, [data_request_callback = move(data_request_callback)](PulseAudioStream&, Bytes buffer, size_t sample_count) {
-            return data_request_callback(buffer, PcmSampleFormat::Float32, sample_count);
-        })));
-
-        // PulseAudio retains the last volume it sets for an application. We want to consistently
-        // start at 100% volume instead.
-        TRY_OR_EXIT_THREAD(internal_state->stream()->set_volume(1.0));
+    // Make sure thread initializes all its state before we move on.
+    Threading::Mutex mut;
+    Threading::ConditionVariable cv(mut);
 
+    // Create the control thread and start it.
+    auto thread = TRY(Threading::Thread::try_create([=, data_request_callback = move(data_request_callback), &cv, &mut]() mutable {
+        mut.lock();
+        {
+            ScopeGuard sg = [&] { cv.signal(); mut.unlock(); };
+            auto context = TRY_OR_EXIT_THREAD(PulseAudioContext::instance());
+            internal_state->set_stream(TRY_OR_EXIT_THREAD(context->create_stream(initial_state, sample_rate, channels, target_latency_ms, [data_request_callback = move(data_request_callback)](PulseAudioStream&, Bytes buffer, size_t sample_count) {
+                return data_request_callback(buffer, PcmSampleFormat::Float32, sample_count);
+            })));
+
+            // PulseAudio retains the last volume it sets for an application. We want to consistently
+            // start at 100% volume instead.
+            TRY_OR_EXIT_THREAD(internal_state->stream()->set_volume(1.0));
+        }
         internal_state->thread_loop();
         return 0;
     },
         "Audio::PlaybackStream"sv));
 
     internal_state->set_thread(thread);
-    thread->start();
+    {
+        Threading::MutexLocker lock(mut);
+        thread->start();
+        cv.wait();
+    }
     thread->detach();
     return playback_stream;
 }
```

Reply:
... though I guess technically my patch *also* has a race where it's possible that we signal the main thread, switch to the main thread, return from the function, and then try to unlock the mutex in the secondary thread :thinking:

Reply:
which means... what. the mutex + cv should be moved to the internal state, and we can make a nice API, `wait_for_initialized()` to call from the main thread, and `signal_initialization_complete()` form the secondary thread?

Reply:
For posterity, the purpose of creating the `PulseAudioStream` within the thread was to do it asynchronously and not block the main thread while still allowing commands to be queued to the `PlaybackStream` implementation. The real solution here was to check for `m_state->stream() == nullptr` and providing a default of `Duration::zero()` if it is not ready yet.